### PR TITLE
Negotiate + SignalR Service Support for the Java client

### PR DESCRIPTION
**Pull request by mikaelm12**
Added support for the negotiation dance between the client and server and also added support for the SignalR service.

Was originally going to make these two different PRs but I really wanted to see the Java client working with the service 😄 
Issue: https://github.com/aspnet/SignalR/issues/2871
and issue: https://github.com/aspnet/SignalR/issues/2862

Need to add tests

### CODE REVIEW

File: clients/java/signalr/build.gradle [hunk 1]
```diff
@@ -18,6 +18,7 @@ dependencies {
     testImplementation group: 'junit', name: 'junit', version: '4.12'
     implementation "org.java-websocket:Java-WebSocket:1.3.8"
     implementation 'com.google.code.gson:gson:2.8.5'
+    compile 'org.apache.httpcomponents:httpclient:4.5.6'
```
Comment (mikaelm12) [inline clients/java/signalr/build.gradle]: New dependency for an implementation of an HttpClient

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 2]
```diff
@@ -22,8 +22,13 @@
     private HubConnectionState connectionState = HubConnectionState.DISCONNECTED;
     private Logger logger;
     private List<Consumer<Exception>> onClosedCallbackList;
+    private boolean skipNegotiate = true;
+    private NegotiateResponse negotiateResponse;
+    private String accessToken;
 
-    public HubConnection(String url, Transport transport, Logger logger){
+    private static int MAX_NEGOTIATE_ATTEMPTS = 5;
```
Comment (mikaelm12) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Just arbitrarily put 5 for the max. I think it's much higher for TS

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 3]
```diff
@@ -145,8 +148,32 @@ public void start() throws Exception {
         if (connectionState != HubConnectionState.DISCONNECTED) {
             return;
         }
+        if (!skipNegotiate) {
+            int negotiateAttempts = 0;
+            do {
+                if(negotiateAttempts > 0 && this.negotiateResponse.accessToken != null){
+                    this.accessToken = negotiateResponse.accessToken;
+                    this.negotiateResponse = Negotiate.processNegotiate(url, this.negotiateResponse.accessToken);
+                    this.url = this.url + "&id=" + negotiateResponse.connectionId + "&access_token=" + this.accessToken;
```
Comment (mikaelm12) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Should I just store the `accessToken`  in the `Authorization` header instead?

Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Yes

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        HttpClient client = HttpClientBuilder.create().build();
+        HttpPost post = new HttpPost(url + "/negotiate");
+        HttpResponse response = client.execute(post);
+        String result = getResponseResult(response);
+        return new NegotiateResponse(result);
+    }
+
+    public static NegotiateResponse processNegotiate(String url, String accessTokenHeader) throws IOException {
+        HttpClient client = HttpClientBuilder.create().build();
+        url = url.substring(0, url.indexOf('?')) + "negotiate" + url.substring(url.indexOf('?'));
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: This assumes the url ends with '/'

Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: Is there some library code you can use instead of trying to correctly parse urls?

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+public class Negotiate {
+
+    public static NegotiateResponse processNegotiate(String url) throws IOException {
+        HttpClient client = HttpClientBuilder.create().build();
+        HttpPost post = new HttpPost(url + "/negotiate");
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: This assumes there is no query string present

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    public Set<String> availableTransports = new HashSet<>();
+    public boolean shouldRedirect;
+    public String redirectUrl;
+    public String accessToken;
+
+
+    public NegotiateResponse(String negotiatePayload) {
+        JsonObject negotiateResponse= jsonParser.parse(negotiatePayload).getAsJsonObject();
+        if (negotiateResponse.has("url")) {
+            this.shouldRedirect = true;
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java]: unneeded, just check if the `redirectUrl` is non-empty

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            this.redirectUrl = negotiateResponse.get("url").getAsString();
+            if (negotiateResponse.has("accessToken")) {
+                this.accessToken = negotiateResponse.get("accessToken").getAsString();
+            }
+            return;
+        }
+        this.connectionId = negotiateResponse.get("connectionId").getAsString();
+        negotiateResponse.get("availableTransports");
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java]: dupe code

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 8]
```diff
@@ -13,15 +16,20 @@
...
     private static final String HTTPS = "https";
     private static final String WS = "ws";
     private static final String WSS = "wss";
+    private static final String ACCESS_TOKEN_PREFIX = "access_token=";
 
     public WebSocketTransport(String url, Logger logger) throws URISyntaxException {
         this.url = formatUrl(url);
         this.logger = logger;
+        if (url.indexOf(ACCESS_TOKEN_PREFIX) > 0){
+            this.accessToken = url.substring(url.indexOf(ACCESS_TOKEN_PREFIX) + ACCESS_TOKEN_PREFIX.length());
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: What is this for?

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java (hunk 7 above)
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java]: This was debugging

File: clients/java/signalr/build.gradle (hunk 1 above)
Comment (mikaelm12) [reply clients/java/signalr/build.gradle]: This has been approved 👍

Comment (anurse) [reply clients/java/signalr/build.gradle]: Looks like [`compile` has been deprecated](https://docs.gradle.org/current/userguide/java_plugin.html#tab:configurations) and `implementation` should be used instead (like the other dependencies)

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java (hunk 2 above)
Comment (anurse) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: It's 100 in .NET: https://github.com/aspnet/SignalR/blob/release/2.2/src/Microsoft.AspNetCore.Http.Connections.Client/HttpConnection.cs#L28

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        String result = getResponseResult(response);
+        return new NegotiateResponse(result);
+    }
+
+    public static NegotiateResponse processNegotiate(String url, String accessTokenHeader) throws IOException {
+        HttpClient client = HttpClientBuilder.create().build();
+        url = url.substring(0, url.indexOf('?')) + "negotiate" + url.substring(url.indexOf('?'));
+        HttpPost post = new HttpPost(url);
+        post.setHeader("Authorization", "Bearer " + accessTokenHeader);
+        HttpResponse response = client.execute(post);
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: It looks like there's an Async flavor of this HttpClient https://hc.apache.org/httpcomponents-asyncclient-ga/httpasyncclient/apidocs/org/apache/http/nio/client/HttpAsyncClient.html
I can update the Async issue to note that we should be using this once we start using `CompletableFuture`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java (hunk 3 above)
Comment (anurse) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: This will just keep appending `&id` and `&access_token` as the client keeps getting redirected. I'm not sure the logic needs to be different based on which attempt we're on. The loop [in the C# client
](https://github.com/aspnet/SignalR/blob/release/2.2/src/Microsoft.AspNetCore.Http.Connections.Client/HttpConnection.cs#L314) doesn't really do that. The initial URL and initial Access Token aren't relevant at all once the negotiate response comes in (https://github.com/aspnet/SignalR/issues/2859 notwithstanding, but we should come back to that)

**Comment by mikaelm12** [comment]
So I was testing this on Android and it turns out that the Apache HttpClient library isn't supported.
https://developer.android.com/about/versions/marshmallow/android-6.0-changes#behavior-apache-http-client

I'm looking into using Squares [OkHttp ](https://github.com/square/okhttp)Library as a replacement.

**Comment by davidfowl** [comment]
OkHttp is the pretty common android replacement. Grpc uses it https://github.com/grpc/grpc-java/tree/master/okhttp

**Comment by mikaelm12** [comment]
>OkHttp is the pretty common android replacement.

Yeah, should have just started there. Hindsight is 20/20 I guess.

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        StringBuffer result = new StringBuffer();
+        String line = "";
+        while ((line = rd.readLine()) != null) {
+            result.append(line);
+        }
+
+        return result.toString();
+    }
+
+    private static String resolveNegotiateUrl(String url){
```
Comment (davidfowl) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: nit: the formatting in this method

File: clients/java/signalr/build.gradle [hunk 11]
```diff
@@ -19,6 +19,7 @@ dependencies {
     implementation "org.java-websocket:Java-WebSocket:1.3.8"
     implementation 'com.google.code.gson:gson:2.8.5'
     implementation 'org.apache.httpcomponents:httpclient:4.5.6'
+    implementation 'com.squareup.okhttp3:okhttp:3.11.0'
```
Comment (davidfowl) [inline clients/java/signalr/build.gradle]: We need to do an OSS request 

cc @Eilon

Comment (mikaelm12) [reply clients/java/signalr/build.gradle]: Already sent an email before updating the PR

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        String result = response.body().string();
+        return new NegotiateResponse(result);
+    }
+
+    private static String resolveNegotiateUrl(String url){
+        String negotiateUrl = "";
+
+        // Check if we have a query string. If we do then we ignore it for now.
+        int queryStringIndex = url.indexOf('?');
+        if(queryStringIndex > 0){
```
Comment (davidfowl) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: you have what I call @anurse virus. The `if(` always touching 😄

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        JsonObject negotiateResponse= jsonParser.parse(negotiatePayload).getAsJsonObject();
+        if (negotiateResponse.has("url")) {
+            this.redirectUrl = negotiateResponse.get("url").getAsString();
+            if (negotiateResponse.has("accessToken")) {
+                this.accessToken = negotiateResponse.get("accessToken").getAsString();
+            }
+            return;
+        }
+        this.connectionId = negotiateResponse.get("connectionId").getAsString();
+        JsonArray transports = ((JsonArray)negotiateResponse.get("availableTransports"));
```
Comment (davidfowl) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/NegotiateResponse.java]: nit: Get rid of the extra parenthesis

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java (hunk 8 above)
Comment (davidfowl) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: This looks like it should be removed.

**Comment by mikaelm12** [comment]
Okay, this now works on Android and I no longer store the accessToken in the url. Moved to headers. And the formatting should be good too

**Comment by mikaelm12** [comment]
Updated

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 14]
```diff
@@ -13,35 +16,47 @@
...
     private URI formatUrl(String url) throws URISyntaxException {
-        if(url.startsWith(HTTPS)){
+        if (url.startsWith(HTTPS)) {
             url = WSS + url.substring(HTTPS.length());
-        }
-        else if(url.startsWith(HTTP)){
+        } else if (url.startsWith(HTTP)) {
             url = WS + url.substring(HTTP.length());
         }
+        if (url.indexOf(ACCESS_TOKEN_PREFIX) > 0) {
```
Comment (davidfowl) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: Why does this exist?

Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: We parse the access token from the url so we can put it in a header when we construct the underlying WebSocketClient. We actually know the accessToken before we even contruct the transport. I can add a ctor that takes headers and get rid of this

Comment (davidfowl) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: That would be much cleaner.

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 15]
```diff
@@ -92,20 +100,23 @@ public HubConnection(String url, Transport transport, Logger logger){
...
-            } catch (URISyntaxException e) {
-                e.printStackTrace();
-            }
-        } else {
+        if (transport != null) {
             this.transport = transport;
         }
     }
 
+    public HubConnection(String url, Transport transport, Logger logger) {
```
Comment (davidfowl) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Why did you add 2 constructors?

Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: For testing. Once I add the skipNegotiate option to the builder I can clean this up

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+public class Negotiate {
+
+    public static NegotiateResponse processNegotiate(String url) throws IOException {
+        return processNegotiate(url, null);
+    }
+
+    public static NegotiateResponse processNegotiate(String url, String accessTokenHeader) throws IOException {
+        url = resolveNegotiateUrl(url);
+        OkHttpClient client = new OkHttpClient();
+        RequestBody body = RequestBody.create(null, new byte[]{});
```
[the remark this answers is not in the document]
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: >What are the create arguments?

`create(MediaType contentType, byte[] content)`

[the remark this answers is not in the document]
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: >Why aren’t we copying all of the headers here?

This seems to be the way we do this in the TS client

**Comment by mikaelm12** [comment]
Ping

File: clients/java/signalr/src/test/java/NegotiateResponseTest.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.microsoft.aspnet.signalr.NegotiateResponse;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class NegotiateResponseTest {
+
+    @Test
+    public void VerifyNegoitateResponse() {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/test/java/NegotiateResponseTest.java]: nit: Negotiate

File: clients/java/signalr/src/test/java/NegotiateResponseTest.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+        assertTrue(negotiateResponse.getAvailableTransports().contains("WebSockets"));
+        assertTrue(negotiateResponse.getAvailableTransports().contains("ServerSentEvents"));
+        assertTrue(negotiateResponse.getAvailableTransports().contains("LongPolling"));
+        assertNull(negotiateResponse.getAccessToken());
+        assertNull(negotiateResponse.getRedirectUrl());
+        assertEquals("bVOiRPG8-6YiJ6d7ZcTOVQ", negotiateResponse.getConnectionId());
+    }
+
+    @Test
+    public void VerifyRedirectNegoitateResponse() {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/test/java/NegotiateResponseTest.java]: nit: Negotiate

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 19]
```diff
@@ -420,18 +477,23 @@ public Subscription on(String target, Action callback) {
...
     public void onClosed(Consumer<Exception> callback) {
-        if (onClosedCallbackList == null){
+        if (onClosedCallbackList == null) {
             onClosedCallbackList = new ArrayList<>();
         }
 
         onClosedCallbackList.add(callback);
     }
+
+    public void setSkipNegotiate(boolean skipNegotiate) {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Remove

**Comment by mikaelm12** [comment]
> There are no tests that check the redirect url behavior.

I wanted to add this but this requires adding infrastructure to allow passing in custom HttpClients for testing. That should happen eventually but not in this PR imo. Trying to stop this PR from growing any more

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 20]
```diff
@@ -72,31 +82,31 @@ public void stop() {
...
-                     onReceive(message);
-                 } catch (Exception e) {
-                     e.printStackTrace();
-                 }
-             }
-
-             @Override
-             public void onClose(int code, String reason, boolean remote) {
+    private WebSocketClient createWebSocket(Map<String, String> headers) {
+        return new WebSocketClient(this.url, headers) {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: Why are you now using `this` when accessing class members?

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        Request request = requestBuilder.build();
+
+        Response response = client.newCall(request).execute();
+        String result = response.body().string();
+        return new NegotiateResponse(result);
+    }
+
+    private static String resolveNegotiateUrl(String url) {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: Add some tests for this please. See https://github.com/aspnet/SignalR/blob/release/2.2/clients/ts/signalr/tests/Common.ts#L17 and https://github.com/aspnet/SignalR/blob/release/2.2/clients/ts/signalr/tests/HttpConnection.test.ts#L220 for some reference

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 22]
```diff
@@ -139,14 +153,47 @@ public HubConnectionState getConnectionState() {
...
      * @throws Exception An error occurred while connecting.
      */
     public void start() throws Exception {
         if (connectionState != HubConnectionState.DISCONNECTED) {
             return;
         }
+        if (!skipNegotiate) {
+            int negotiateAttempts = 0;
+            do {
+                this.accessToken = (this.negotiateResponse == null) ? null : this.negotiateResponse.getAccessToken();
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: this this this

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (anurse) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: It's a little hidden, but we do copy all the headers in the JavaScript client:

https://github.com/aspnet/SignalR/blob/201c058b9379b584317b912a4928a83ad7b569bf/clients/ts/signalr/src/HttpClient.ts#L184-L190

**Comment by anurse** [comment]
Formatting changes made a bunch of noise in HubConnection unfortunately :(

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java (hunk 21 above)
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: Good point. 👍

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java (hunk 20 above)
Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: I actually want to move towards using `this`. Since it's not the convention to use an underscore prefix I've started liking the idea of using `this`

**Comment by mikaelm12** [comment]
Final reviews?

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+        }
+
+        Request request = requestBuilder.build();
+
+        Response response = client.newCall(request).execute();
+        String result = response.body().string();
+        return new NegotiateResponse(result);
+    }
+
+    public static String resolveNegotiateUrl(String url) {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/Negotiate.java]: We'll need to look at the public API later
